>  in the "src\Http\WebUtilities\src" directory.

Ah, didn't realize the usage over there. Thx.

> some compile errors.

E.g.
```
/home/vsts/work/1/s/src/Shared/WebEncoders/WebEncoders.cs(213,40): error CS1503: Argument 1: cannot convert from 'System.Span<char>' to 'char*' [/home/vsts/work/1/s/src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj]
/home/vsts/work/1/s/src/Shared/WebEncoders/WebEncoders.cs(333,21): error CS0117: 'Convert' does not contain a definition for 'TryToBase64Chars' [/home/vsts/work/1/s/src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj]
```

I believed it's .NET Core 3.0 only...proved wrong https://github.com/aspnet/AspNetCore/blob/0b340bd70cfe41200da593c08f4a88c6fe9efa74/src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj#L4

Will fix this in the next commit.

--- a/src/Http/WebUtilities/ref/Microsoft.AspNetCore.WebUtilities.netcoreapp3.0.cs
+++ b/src/Http/WebUtilities/ref/Microsoft.AspNetCore.WebUtilities.netcoreapp3.0.cs
@@ -230,6 +230,7 @@ namespace Microsoft.AspNetCore.WebUtilities
         public static byte[] Base64UrlDecode(string input) { throw null; }
         public static byte[] Base64UrlDecode(string input, int offset, char[] buffer, int bufferOffset, int count) { throw null; }
         public static byte[] Base64UrlDecode(string input, int offset, int count) { throw null; }
+        public static string Base64UrlEncode(System.ReadOnlySpan<byte> input) { throw null; }
         public static string Base64UrlEncode(byte[] input) { throw null; }
         public static int Base64UrlEncode(byte[] input, int offset, char[] output, int outputOffset, int count) { throw null; }
         public static string Base64UrlEncode(byte[] input, int offset, int count) { throw null; }

--- a/src/Http/WebUtilities/ref/Microsoft.AspNetCore.WebUtilities.netcoreapp3.0.cs
+++ b/src/Http/WebUtilities/ref/Microsoft.AspNetCore.WebUtilities.netcoreapp3.0.cs
@@ -230,10 +230,10 @@ namespace Microsoft.AspNetCore.WebUtilities
         public static byte[] Base64UrlDecode(string input) { throw null; }
         public static byte[] Base64UrlDecode(string input, int offset, char[] buffer, int bufferOffset, int count) { throw null; }
         public static byte[] Base64UrlDecode(string input, int offset, int count) { throw null; }
-        public static string Base64UrlEncode(System.ReadOnlySpan<byte> input) { throw null; }
         public static string Base64UrlEncode(byte[] input) { throw null; }
         public static int Base64UrlEncode(byte[] input, int offset, char[] output, int outputOffset, int count) { throw null; }
         public static string Base64UrlEncode(byte[] input, int offset, int count) { throw null; }
+        public static string Base64UrlEncode(System.ReadOnlySpan<byte> input) { throw null; }
         public static int GetArraySizeRequiredToDecode(int count) { throw null; }
         public static int GetArraySizeRequiredToEncode(int count) { throw null; }
     }

--- a/src/Shared/WebEncoders/WebEncoders.cs
+++ b/src/Shared/WebEncoders/WebEncoders.cs
@@ -195,36 +195,6 @@ namespace Microsoft.Extensions.Internal
         /// </summary>
         /// <param name="input">The binary input to encode.</param>
         /// <returns>The base64url-encoded form of <paramref name="input"/>.</returns>
-        public static string Base64UrlEncode(ReadOnlySpan<byte> input)
-        {
-            if (input.IsEmpty)
-            {
-                return string.Empty;
-            }
-
-            int bufferSize = GetArraySizeRequiredToEncode(input.Length);
-
-            char[] bufferToReturnToPool = null;
-            Span<char> buffer = bufferSize <= 128
-                ? stackalloc char[bufferSize]
-                : bufferToReturnToPool = ArrayPool<char>.Shared.Rent(bufferSize);
-
-            var numBase64Chars = Base64UrlEncode(input, buffer);
-            var base64Url = new string(buffer.Slice(0, numBase64Chars));
-
-            if (bufferToReturnToPool != null)
-            {
-                ArrayPool<char>.Shared.Return(bufferToReturnToPool);
-            }
-
-            return base64Url;
-        }
-
-        /// <summary>
-        /// Encodes <paramref name="input"/> using base64url encoding.
-        /// </summary>
-        /// <param name="input">The binary input to encode.</param>
-        /// <returns>The base64url-encoded form of <paramref name="input"/>.</returns>
         public static string Base64UrlEncode(byte[] input)
         {
             if (input == null)
@@ -251,7 +221,20 @@ namespace Microsoft.Extensions.Internal
 
             ValidateParameters(input.Length, nameof(input), offset, count);
 
+#if NETCOREAPP3_0
             return Base64UrlEncode(input.AsSpan(offset, count));
+#else
+            // Special-case empty input
+            if (count == 0)
+            {
+                return string.Empty;
+            }
+
+            var buffer = new char[GetArraySizeRequiredToEncode(count)];
+            var numBase64Chars = Base64UrlEncode(input, offset, buffer, outputOffset: 0, count: count);
+
+            return new string(buffer, startIndex: 0, length: numBase64Chars);
+#endif
         }
 
         /// <summary>
@@ -302,7 +285,41 @@ namespace Microsoft.Extensions.Internal
                     nameof(count));
             }
 
+#if NETCOREAPP3_0
             return Base64UrlEncode(input.AsSpan(offset, count), output.AsSpan(outputOffset));
+#else
+            // Special-case empty input.
+            if (count == 0)
+            {
+                return 0;
+            }
+
+            // Use base64url encoding with no padding characters. See RFC 4648, Sec. 5.
+
+            // Start with default Base64 encoding.
+            var numBase64Chars = Convert.ToBase64CharArray(input, offset, count, output, outputOffset);
+
+            // Fix up '+' -> '-' and '/' -> '_'. Drop padding characters.
+            for (var i = outputOffset; i - outputOffset < numBase64Chars; i++)
+            {
+                var ch = output[i];
+                if (ch == '+')
+                {
+                    output[i] = '-';
+                }
+                else if (ch == '/')
+                {
+                    output[i] = '_';
+                }
+                else if (ch == '=')
+                {
+                    // We've reached a padding character; truncate the remainder.
+                    return i - outputOffset;
+                }
+            }
+
+            return numBase64Chars;
+#endif
         }
 
         /// <summary>
@@ -317,6 +334,37 @@ namespace Microsoft.Extensions.Internal
         {
             var numWholeOrPartialInputBlocks = checked(count + 2) / 3;
             return checked(numWholeOrPartialInputBlocks * 4);
+        }
+
+#if NETCOREAPP3_0
+        /// <summary>
+        /// Encodes <paramref name="input"/> using base64url encoding.
+        /// </summary>
+        /// <param name="input">The binary input to encode.</param>
+        /// <returns>The base64url-encoded form of <paramref name="input"/>.</returns>
+        public static string Base64UrlEncode(ReadOnlySpan<byte> input)
+        {
+            if (input.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            int bufferSize = GetArraySizeRequiredToEncode(input.Length);
+
+            char[] bufferToReturnToPool = null;
+            Span<char> buffer = bufferSize <= 128
+                ? stackalloc char[bufferSize]
+                : bufferToReturnToPool = ArrayPool<char>.Shared.Rent(bufferSize);
+
+            var numBase64Chars = Base64UrlEncode(input, buffer);
+            var base64Url = new string(buffer.Slice(0, numBase64Chars));
+
+            if (bufferToReturnToPool != null)
+            {
+                ArrayPool<char>.Shared.Return(bufferToReturnToPool);
+            }
+
+            return base64Url;
         }
 
         private static int Base64UrlEncode(ReadOnlySpan<byte> input, Span<char> output)
@@ -353,6 +401,7 @@ namespace Microsoft.Extensions.Internal
 
             return charsWritten;
         }
+#endif
 
         private static int GetNumBase64PaddingCharsInString(string str)
         {

